lib/BloqadeExpr/src/types.jl

```diff
@@ -281,6 +281,66 @@ end
 
 SumOfZ(n::Int) = SumOfZ(n, 1)
 
+
+RabiTypes = Union{Nothing,SumOfX,SumOfXPhase}
+DetuningTypes = Union{Nothing,SumOfN}
+
+struct RydbergHamiltonian{RabiType <: RabiTypes, DetuningType <: DetuningTypes} <: AbstractTerm
+    rydberg_term::RydInteract
+    rabi_term::RabiType
+    detuning_term::DetuningType
+end
+
+@inline add_terms(h::RydbergHamiltonian{Nothing,Nothing}) = YaoBlocks.Optimise.simplify(h.rydberg_term)
+@inline add_terms(h::RydbergHamiltonian{Nothing,SumOfN}) = YaoBlocks.Optimise.simplify(h.rydberg_term - h.detuning_term)
+@inline add_terms(h::RydbergHamiltonian{<:Union{SumOfX,SumOfXPhase},Nothing}) = YaoBlocks.Optimise.simplify(h.rydberg_term + h.rabi_term)
+@inline add_terms(h::RydbergHamiltonian) = YaoBlocks.Optimise.simplify(h.rydberg_term + h.rabi_term - h.detuning_term)
+
+
+get_rydberg_params(h::RydbergHamiltonian{Nothing,Nothing}) = (C=h.rydberg_term.C,
```

Comment:
hmm, maybe we should do something like following directly? and just overload `to_basictypes`

```julia
struct RydbergHamiltonian
   C
   phi
   omega
   delta
end
```

Reply:
I think if `get_rydberg_params` is necessary then directly having the parameters inside the `struct` would be easier. We don't have to shoe-hole something like this.

Also for this type of combination, it is better to not use multiple dispatch, but use `if else` style dispatch, you can use pattern matching for syntax sugar. https://github.com/thautwarm/MLStyle.jl We already depend on this package for `@match` and `@switch`